crates/env/src/engine/on_chain/impls.rs

```diff
@@ -517,8 +517,12 @@ impl TypedEnvBackend for EnvInstance {
             }
             Err(ext::Error::CalleeReverted) => {
                 // We don't wrap manually with an extra `Err` like we do in the `Ok` case since the
-                // buffer already comes back in the form of `Err(LangError)`
-                let out = scale::Decode::decode(&mut &out_return_value[..])?;
+                // buffer already comes back in the form of `Err(LangError)` (assuming it's encoded
+                // by the ink! codegen and not the contract).
+                let out = ink_primitives::ConstructorResult::<E::AccountId>::decode(
+                    &mut &out_return_value[..],
+                )?;
+                assert!(out.is_err(), "The callee reverted, but did not encode an error in the output buffer.");
```

Comment:
I wonder if this would itself be considered a type of `LangError`. e.g. `LangError::InvalidConstructorRevertedReturnValue` (obvs less verbose)

Reply:
In fact this looks like this `assert` would fail if a fallible constructor returns an error: https://github.com/paritytech/ink/blob/507d01dcb2848c3deb16691bac9ec0c15bcda900/crates/ink/codegen/src/generator/dispatch.rs#L635-L643

Because it will revert, but the outer `Ok` is hardcoded.

Reply:
Thanks for bringing up the case for fallible constructors. 

Even without the `assert!` this would've failed to decode. This is because with fallible
constructors we need to decode into a `ConstructorResult<Result<(), E>>`,
instead of a `ConstructorResult<()>` or `ConstructorResult<AccountId>`.

I need to think a bit more about what changes will be required.